### PR TITLE
Added FileNotFound error code according to new Mavlink FTP specification

### PR DESCRIPTION
**Pull request by MatejFranceskin**
Update to Mavlink FTP protocol added a new error code for FileNotFound error.
https://github.com/mavlink/mavlink-devguide/pull/199

Until now a errno was set to ENOENT and had to be interpreted as such in MavlinkFTP client.

Code change is aligned and tested with MAVSDK MavlinkFTP module and its integration tests.

### CODE REVIEW

**Comment by dagar** [comment]
Mavlink FTP tests failing on Jenkins.
http://ci.px4.io:8080/blue/organizations/jenkins/PX4%2FFirmware/detail/PR-12976/1/pipeline

image

**Comment by LorenzMeier** [comment]
@MatejFranceskin You need to update the test as well or fix the implementation. If the test needs to be updated then the backwards compatibility might have been broken, which is a hint that the change needs to be fixed. We can't upgrade all QGC / SDK version in parallel.

**Comment by MatejFranceskin** [comment]
One test actually pointed out to a real problem which is fixed now. Other tests have been changed.

QGC currently doesn't check the errno return codes and will behave the same regardless of this PR. When this is merged I will make a PR also for QGC so that it will display new error message.

MAVSDK MavlinkFTP module will work with older and newer firmwares.

File: src/modules/mavlink/mavlink_tests/mavlink_ftp_test.cpp [hunk 1]
```diff
@@ -753,20 +759,22 @@ bool MavlinkFtpTest::_removefile_test()
...
-		{ _unittest_microsd_file,	true },
-		{ _unittest_microsd_file,	false },
+		{ _unittest_microsd_dir,	false, 2, MavlinkFTP::kErrFailErrno },
+		{ _unittest_microsd_file,	true,  0, MavlinkFTP::kErrNone },
+		{ _unittest_microsd_file,	false, 1, MavlinkFTP::kErrFileNotFound },
 	};
 
 	ut_compare("mkdir failed", ::mkdir(_unittest_microsd_dir, S_IRWXU | S_IRWXG | S_IRWXO), 0);
-	ut_assert("open failed", (fd = ::open(_unittest_microsd_file, O_CREAT | O_EXCL, S_IRWXU | S_IRWXG | S_IRWXO)) != -1);
+	ut_assert("open failed", (fd = ::open(_unittest_microsd_file, O_CREAT | O_EXCL, S_IRWXU | S_IRWXG | S_IRWXO)) != 0);
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_tests/mavlink_ftp_test.cpp]: Why did you change this?

Comment (MatejFranceskin) [reply src/modules/mavlink/mavlink_tests/mavlink_ftp_test.cpp]: Mistake. Fixed it.

**Comment by dagar** [comment]
@MatejFranceskin does `make tests` work locally?

**Comment by hamishwillee** [comment]
@julianoes Can you merge this if it is OK?

**Comment by julianoes** [comment]
Rebased and force pushed, let's see.

**Comment by julianoes** [comment]
I'm puzzled. The tests pass locally for me as well as in docker. This must be some permission issue that we can't write files in the working directory.

**Comment by bkueng** [comment]
Can you add more instrumentation? The result shows `Compare failed: Didn't get Nak back` indicating a test that should fail did not.

File: src/modules/mavlink/mavlink_ftp.h [hunk 2]
```diff
@@ -45,6 +45,8 @@
 
 #include "mavlink_bridge_header.h"
 
+#define MAVLINK_FTP_DEBUG 1
```
Comment (bkueng) [inline src/modules/mavlink/mavlink_ftp.h]: Is this intended to be set to 1?

Comment (julianoes) [reply src/modules/mavlink/mavlink_ftp.h]: Oops, thx.

**Comment by bkueng** [comment]
CI still fails on MacOS.

**Comment by LorenzMeier** [comment]
Finally fixed - merging.

**Comment by julianoes** [comment]
@LorenzMeier thanks for merging but you managed to squash (overwrite) 22h hours of my time and 11 commits with valuable diff information :astonished:. I don't care too much about my name being in the history but more about the fact that I wrote down why and how I changed things for the next confused reader. And squashing as a final step is ok but not even having it in the GitHub PR here is bad.

**Comment by LorenzMeier** [comment]
Fair ask regarding the commit messages. From the summary lines it did sound like CI debugging, that might have been the wrong call here, sorry! I’m wondering if some of that information would have rather belonged into code or a dedicated CI doc?

I believe that most people only will look at a commit message if they dont understand the change and some of the information you have in there would probably benefit anyone extending these tests in the future.
